docs/rfcs/2025-08-11-channels-forwarding.md

```diff
@@ -0,0 +1,188 @@
+- For a given pair of members only a single admin forwards messages - inviting admin (host) forwards messages between its invitees and members introduced to them. It means:
+  - This admin can be a single point of failure and/or can cause arbitrary delays in message delivery based on its availability when forwarding messages between these member pairs.
+  - Member pairs fully trust their single forwarding admin in forwarded content.
+
+Not limitations of protocol, but other weak points of current implementation:
+- Forward operations are synchronous to message reception (by forwarding admin), so they're not resumable on failure.
+- Forward operations are carried out as "group sends" to all required members (invitees/introduced members in respect to sending member), number of which is expected to grow very large in channels and can potentially have a big load on the database.
+
+## Solution
+
+Chat relays will serve as forwarding agents instead of inviting admins, with following modifications to protocol and implementation.
+
+### Forwarding not limited to establishing connection
+
+- In channels members will not connect directly to other members and channel owners.
+- Chat relays will not take into consideration status of introductions.
+- Chat relays will forward messages between owners and all members (from members to owners, for example, for reactions, comments).
+
+### All chat relays forward all messages
+
+- Instead of inviting admins all chat relays will forward all messages.
+- Members will deduplicate messages (already implemented).
+- Members should highlight differences in deduplicated messages. This would solve trust issue.
+  - Question: If messages are to be signed by owners, what differences can chat relays introduce? Does this point in initial doc imply owner sending different versions of message to chat relays? Losses in delivery? Answer: Not all messages will be signed.
+
+#### Highlighting deduplicated messages differences
+
+Currently we simply ignore duplicate messages - see createNewRcvMessage.
+
+Some options on how to highlight difference in messages from chat relays:
+1. Show only the fact that there is difference.
+    - Persist duplicate error (SEDuplicateGroupMessage) as flag on chat item to then display warning sign in UI.
+2. Save message hashes as some additional entity linked to message or chat item for each chat relay.
+    - If any new hash differs, set flag on chat item for warning sign in UI.
+    - Add ability to load additional info (with other chat item info via APIGetChatItemInfo) showing which relays sent different hashes (e.g. 1 differs from 2 others - show 1 as warning, all differ - show all as warning).
+3. As previous, but save content. Possibly save content additionally to hashes, for faster comparison.
+    - Saving content means converting duplicate messages to chat item content, or only saving text, but latter will fail to show difference in files.
```

Comment:
why can't we have them as usual chat items that would show next to duplicates? Or make a special "different version of another chat item" as a special reply (to utilize reply machinery)?

As they would have the same timestamp (would they?), they would show next to original. If timestamp is different (probably), we could make it the same. It won't work for replies though.

In this case we could simply have a flag "different message from X relay" so users can see which relay?